RealmSwift/SwiftUI.swift

```diff
@@ -643,6 +653,390 @@ extension EnvironmentValues {
+            asyncOpenForUser(currentUser, partitionValue: partitionValue, configuration: configuration)
+        } else {
+            storage.asyncOpenState = .waitingForUser
+            app.objectWillChange.sink { [self] app in
+                if let currentUser = app.currentUser {
+                    asyncOpenForUser(currentUser, partitionValue: partitionValue, configuration: configuration)
+                }
+            }.store(in: &storage.cancellables)
+        }
+    }
+
+    private func asyncOpenForUser(_ user: User, partitionValue: AnyBSON, configuration: Realm.Configuration) {
+        storage.asyncOpenForUser(user, partitionValue: partitionValue, configuration: configuration)
+            .sink { completion in
+                if case .failure(let error) = completion {
+                    self.storage.asyncOpenState = .error(error)
+                }
+                cancel()
+            } receiveValue: { realm in
+                self.storage.asyncOpenState = .open(realm)
+            }.store(in: &storage.cancellables)
+    }
+
+    /**
+     A Publisher for `AsyncOpenState`, emits a state each time the asyncOpen state changes.
+     */
+    public var projectedValue: Published<AsyncOpenState>.Publisher {
+        return storage.$asyncOpenState
+    }
+
+    /// :nodoc:
+    public var wrappedValue: AsyncOpenState {
+        storage.asyncOpenState
+    }
+
+    /**
+     This will cancel any notification from the property wrapper states
+     */
+    public func cancel() {
+        storage.cancellables.forEach { $0.cancel() }
+        storage.cancellables = []
+    }
+
+    /**
+     Initialize the property wrapper
+     - parameter appId: The unique identifier of your Realm app, if empty or `nil` will try to retrieve latest singular cached app.
+     - parameter partitionValue: The `BSON` value the Realm is partitioned on.
+     - parameter configuration: The `Realm.Configuration` used when creating the Realm,
+                 user's sync configuration for the given partition value will be set as the `syncConfiguration`,
+                 if empty the configuration is set to the `defaultConfiguration`
+     - parameter timeout: The maximum number of milliseconds to allow for a connection to
+                 become fully established., if empty or `nil` no connection timeout is set.
+     */
+    public init<T: BSON>(appId: String? = nil, partitionValue: T, configuration: Realm.Configuration = Realm.Configuration.defaultConfiguration, timeout: UInt? = nil) {
+        var app: App
+        if let appId = appId {
+            app = App(id: appId)
+        } else if RLMApp.apps().allKeys.count == 1, // Check if there is a singular cached app
+            let cachedAppId = RLMApp.apps().allKeys.first as? String {
+            app = App(id: cachedAppId)
+        } else {
+            throwRealmException("There is no appId, either provided by the user on the property wrapper or 'any/more than 1' cached RLMApp")
```

Comment:
I would break this up into different error messages for each case to better guide the user.